### ISSUE TITLE
No timeout when mounting ZFS - system will never finish booting

### ISSUE THREAD

**Issue by Malvineous** [archzfs/archzfs#345]
I'm not sure if this is an Arch-specific issue as I guess it's systemd related, but I've discovered that if you can't mount a ZFS filesystem at boot for some reason, the system never finishes booting and is unusable.  You have to boot from a USB stick or something to repair the install, otherwise you end up like this, where the system has been stuck for an hour with no signs of continuing:

![zfs-slowboot](https://user-images.githubusercontent.com/171656/86922147-909d9b00-c16f-11ea-80a8-23437d94b35b.png)

While this might be understandable if the root filesystem was on ZFS, on this machine ZFS is only handling bulk storage and the system can boot just fine without any ZFS filesystems being mounted.

Since you can't even SSH into the machine or log in via a console when it's in this state, I think there should be *some* limit by default (even if it's fairly high) just so the system can finish booting and be recovered, rather than being frozen forever.

**Comment by sam-lunt** [comment]
This isn't an Arch-specific issue. The `zfs-mount.service` unit doesn't have a timeout provided (I don't think that it makes sense to add one by default, but that's beside the point). If _you_ would like to add one, you can easily do so with a [Systemd drop-in file](https://wiki.archlinux.org/index.php/Systemd#Drop-in_files).

Also, if you are using mkinitcpio to generate your initramfs, then you can pass the kernel command-line parameter `break=postmount` (`break=premount` is also an option, but wouldn't help here) in order to enter an interactive shell after the root filesystem is mounted. You can then disable `zfs-mount.service` or correct the ZFS cache (`zfs-mount` is just running `zfs mount -a`).

**Comment by Malvineous** [comment]
Fair enough, although when you say you don't think it makes sense to add a timeout by default, can you let me know a scenario where it is undesirable that it drops to single-user mode and it's better that it simply locks up forever in the middle of the boot process with no way out other than hard-resetting the machine and facing possible filesystem corruption from non-ZFS volumes?  I can't think of any situation where having the system frozen permanently is better than giving you the option of logging in to fix the issue.

In my case the problem ended up being user error, as I had a mistake in a mount point so two mount points were depending on each other.  But given that this machine takes around six minutes to reboot, even rebooting and adding a kernel parameter is a bit of an ordeal.  Having it drop to single user mode and letting me log in after say 10 minutes would be a much more sensible default in my opinion, as taking more than 10 minutes to mount ZFS volumes doesn't seem like a typical situation for most users to me.

Happy to raise this issue upstream but I couldn't work out where the systemd unit files were coming from.

**Comment by sam-lunt** [comment]
The systemd file is owned by `zfs-utils`:
```
$ pacman -Qo /usr/lib/systemd/system/zfs-mount.service
/usr/lib/systemd/system/zfs-mount.service is owned by zfs-utils 0.8.4-1
```

I'm hardly an expert, but the reason I said it wouldn't make sense to have a timeout is that because the boot times can vary so much between systems, it would be hard to choose a timeout value that works for everyone. You actually provide a great example yourself: I would say that 6 minutes is a very long boot-up time. Many systems will boot in under a minute, so a timeout of 10 minutes would be quite long. It would effectively be like having no timeout at all, since most people would give up after a minute or so. In fact, a timeout of 10 minutes would probably be worse than no timeout at all, since someone might see that timeout and decide to wait 10 minutes instead of just rebooting earlier.

The other thing that could be an issue is if that mount step is responsible for mounting something like `/var/log`. If that isn't mounted, then the system might try to create directories on the dataset that contains `/` or `/var`. When you then fix the issue that kept `/var/log` from mounting, you'll have to deal with the fact that two separate log directories. If it's something like `/var` that isn't mounted, then you might have even more problems, since there will be a lot of files that are missing. Now instead of having one problem (you can't mount a dataset), you have two problems (you can't mount the dataset and your system is complaining because it thinks you erased `/var`)

That said, I'm sure you would get a better answer if you made an issue with the ZFSonLinux project. Maybe they would disagree with me, and the lack of a timeout is just because no one has suggested it yet.

**Comment by dngray** [comment]
![CaptureScreen](https://user-images.githubusercontent.com/48640805/222104136-6d0e6d2d-ce96-44c5-9f50-f587f0d85537.jpeg)

So I am finding this exact issue.

My root is a btrfs disk, and my ZFS keys are stored in `/etc/zfs/keys`. The issue is that zfs tries to mount the datasets before I have a chance to put in my LUKS password and unlock my btrfs root.

Any ideas how I can solve this?

**Comment by stevleibelt** [comment]
So overall, what we could do is to inform users and gave them a simple way how to set the specific timeout value in a self created `/etc/systemd/system/zfs-mount.service.d/override.conf`?

**Comment by Malvineous** [comment]
Actually that's a very good point.  You could just put an override in like that.  But I think it would be good to document that somewhere because it's not obvious if you're not familiar with systemd.

**Comment by stevleibelt** [comment]
@minextu do you see a good point where one of us could create a pull request with a documentation of this information inside?

**Comment by minextu** [comment]
If it's documenting something that already works, the wiki would be a good place for it I think. ArchZFS specific issues are good in the github wiki (https://github.com/archzfs/archzfs/wiki), but for zfs in general the archlinux wiki is probably a better place (https://wiki.archlinux.org/title/ZFS#Troubleshooting)
